src/_common/content/content-viewer/content-viewer.ts

```diff
@@ -97,49 +98,53 @@ export default class AppContentViewer extends Vue implements ContentOwner, Light
 		this.data = content;
 		this.hydrator = new ContentHydrator(content.hydration);
 
-		if (this.isInview) {
-			this.mediaItems = this.getMediaItems(this.data);
-			this.mediaItemsNeedRefreshed = false;
-			return;
-		}
-
-		this.mediaItemsNeedRefreshed = true;
+		this.mediaItemsResolved = false;
+		this._resolveMediaItems();
 	}
 
 	onScrollInview() {
 		this.isInview = true;
-		if (this.mediaItemsNeedRefreshed && this.data) {
-			this.mediaItems = this.getMediaItems(this.data);
-			this.mediaItemsNeedRefreshed = false;
-		}
+		this._resolveMediaItems();
 	}
 
 	onScrollOutview() {
 		this.isInview = false;
 	}
 
-	getMediaItems(data: ContentDocument) {
-		const items = data.getChildrenByType('mediaItem');
+	private _resolveMediaItems() {
+		if (this.mediaItemsResolved) {
+			return;
+		}
+
+		const items = this.data?.getChildrenByType('mediaItem') || [];
 		if (!items.length) {
-			return [];
+			this.mediaItems = [];
+			return;
 		}
 
+		if (this.isInview || GJ_IS_SSR) {
+			this.mediaItems = this.getMediaItems(this.owner, items);
+			this.mediaItemsResolved = true;
+		}
+	}
+
+	getMediaItems(owner: ContentOwner, items: ContentObject[]) {
 		const processedItems: MediaItem[] = [];
 
 		items.forEach(item => {
 			if (item.attrs.href) {
 				return;
 			}
 
-			let _mediaItem: MediaItem | null = null;
-			this.owner.getHydrator().useData('media-item-id', item.attrs.id.toString(), data => {
+			let processedItem: MediaItem | null = null;
+			owner.getHydrator().useData('media-item-id', item.attrs.id.toString(), data => {
 				if (data) {
-					_mediaItem = new MediaItem(data);
+					processedItem = new MediaItem(data);
 				}
 			});
 
-			if (_mediaItem) {
-				processedItems.push(_mediaItem);
+			if (processedItem) {
+				processedItems.push(processedItem);
```

Comment:
This way of retruning processedItems wont work. If hydrator doesn't have the data it'll add the items to the array only after the function returns it which is unexpected.

Since you need the getMediaItems to be syncronous for SSR I don't think you can promisify it..
Lets talk about this.